src/actions/package/repository.rs

```diff
@@ -0,0 +1,102 @@
+use crate::actions::Action;
+use crate::contexts::Contexts;
+use crate::manifests::Manifest;
+use crate::steps::Step;
+use serde::{Deserialize, Serialize};
+use std::ops::Deref;
+use tracing::{error, span};
+
+use super::providers::PackageProviders;
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub struct PackageRepository {
+    #[serde(alias = "url")]
+    pub name: String,
+
+    pub key: Option<RepositoryKey>,
+
+    #[serde(default)]
+    pub provider: PackageProviders,
+}
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub struct RepositoryKey {
+    pub url: String,
+    pub name: Option<String>,
+    pub key: Option<String>,
+    pub fingerprint: Option<String>,
+}
+
+impl Action for PackageRepository {
+    fn plan(&self, _manifest: &Manifest, _context: &Contexts) -> Vec<Step> {
+        let box_provider = self.provider.clone().get_provider();
+        let provider = box_provider.deref();
+
+        let span = span!(
+            tracing::Level::INFO,
+            "package.repository",
+            provider = provider.name()
+        )
+        .entered();
+
+        let mut atoms: Vec<Step> = vec![];
+
+        // If the provider isn't available, see if we can bootstrap it
+        if !provider.available() {
+            if provider.bootstrap().is_empty() {
+                error!(
+                    "Package Provider, {}, isn't available. Skipping action",
+                    provider.name()
+                );
+                return vec![];
+            }
+
+            atoms.append(&mut provider.bootstrap());
+        }
+
+        if !provider.has_repository(&self) {
+            atoms.append(&mut provider.add_repository(&self));
+        }
+
+        span.exit();
+
+        atoms
+    }
+}
+
+// #[cfg(test)]
```

Comment:
Do you want to throw it away or keep it?

Reply:
I'll come back to them 😅